images/src/org/intellij/images/util/imageio/svg/SvgImageReader.kt

```diff
@@ -0,0 +1,144 @@
+// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
+package org.intellij.images.util.imageio.svg
+
+import com.github.weisj.jsvg.attributes.font.SVGFont
+import com.github.weisj.jsvg.geometry.size.Length
+import com.github.weisj.jsvg.nodes.SVG
+import com.intellij.ui.paint.PaintUtil
+import com.intellij.ui.scale.ScaleContext
+import com.intellij.ui.svg.createJSvgDocument
+import com.intellij.util.ui.ImageUtil
+import java.awt.image.BufferedImage
+import java.io.IOException
+import java.io.InputStream
+import javax.imageio.ImageReadParam
+import javax.imageio.ImageReader
+import javax.imageio.ImageTypeSpecifier
+import javax.imageio.stream.ImageInputStream
+import kotlin.math.max
+
+class SvgImageReader(svgImageReaderSpi: SvgImageReaderSpi) : ImageReader(svgImageReaderSpi) {
+  private var height: Length? = null
+  private var width: Length? = null
+  private var jSvgDocument: SVG? = null
+  override fun setInput(input: Any?, seekForwardOnly: Boolean, ignoreMetadata: Boolean) {
+    super.setInput(input, seekForwardOnly, ignoreMetadata)
+    reset()
+  }
+
+  override fun getWidth(imageIndex: Int): Int {
+    loadInfoIfNeeded()
+    return width?.raw()?.toInt() ?: throw IOException("SVG document not loaded")
+  }
+
+  override fun getHeight(imageIndex: Int): Int {
+    loadInfoIfNeeded()
+    return height?.raw()?.toInt() ?: throw IOException("SVG document not loaded")
+  }
+
+  @Throws(IOException::class)
+  private fun loadInfoIfNeeded() {
+    val input = input
+    if (jSvgDocument == null && input is ImageInputStream) {
+
+      try {
+        // Not using ByteArray to avoid potential humongous allocation
+        ImageInputStreamAdapter(input).buffered().use {
+          jSvgDocument = createJSvgDocument(it).also { svg ->
+            width = svg.width
+            height = svg.height
+          }
+        }
+      } catch (e: IOException) {
+        // could not read the SVG document
+      }
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun read(imageIndex: Int, param: ImageReadParam?): BufferedImage {
+    loadInfoIfNeeded()
+    jSvgDocument ?: throw IOException("SVG document not loaded")
+
+    val sourceRenderSize = param?.sourceRenderSize
+
+    val width =  sourceRenderSize?.width?.toDouble() ?: width!!.raw().toDouble()
+    val height = sourceRenderSize?.height?.toDouble() ?: height!!.raw().toDouble()
```

Comment:
**thought:** I have two questions on the size of the image :

1. What would be the right scale, given this code missed the context. This can be problematic in multi monitor setup.

2. If the document or the params have large dimensions this could create a huge `BufferedImage`. Is there any specific guideline there? Note, the same could happen with other binary images, if their metadata indicate large dimensions.

Reply:
Also topic (2) may have an aggravating factor with the _scale_, because I think it could doubles the dimensions on HiDPI monitors.